File: frontend-web/verqo-web/src/app/core/validators/gstin.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Client-side GSTIN format check — mirrors Verqo.Application.Kyc.GstinValidator
 * on the .NET side exactly, for instant feedback. GSTIN is optional at Client
 * registration, so this validator (like the .NET one) only rejects a value
 * that's actually present but malformed; it never enforces `required`. The
 * API re-validates (never trust the client) and, same as the .NET side,
 * deliberately doesn't recompute the mod-36 check digit — see
 * GstinValidator's remarks for why.
 */
const GSTIN_FORMAT = /^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$/;
const VALID_PAN_HOLDER_TYPE_CODES = new Set(['P', 'C', 'H', 'A', 'B', 'G', 'J', 'L', 'F', 'T']);

export function gstinValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    const raw = (control.value ?? '').toString().trim().toUpperCase();
    if (!raw) return null; // GSTIN is optional — nothing to validate

    if (!GSTIN_FORMAT.test(raw)) {
      return {
        gstinFormat:
          "GSTIN must be 15 characters: a 2-digit state code, the business's 10-character PAN, an entity number, 'Z', and a check digit.",
      };
    }
    const stateCode = Number(raw.slice(0, 2));
    if (stateCode < 1 || stateCode > 38) {
      return { gstinFormat: 'The first two digits of a GSTIN must be a valid state code (01–38).' };
    }
    if (!VALID_PAN_HOLDER_TYPE_CODES.has(raw[5])) {
      return { gstinFormat: 'The PAN embedded in the GSTIN (characters 3–12) is not validly formatted.' };
    }
    return null;
  };
}

File: backend/src/Verqo.Application/Kyc/GstinValidator.cs
using System.Text.RegularExpressions;

namespace Verqo.Application.Kyc;

/// <summary>
/// Offline GSTIN (Goods and Services Tax Identification Number) structural
/// validation, in the same spirit as <see cref="PanValidator"/>: what can
/// legitimately be checked without calling anyone. A GSTIN is 15 characters
/// — a 2-digit state code, the business's 10-character PAN, a 1-character
/// entity/registration number, the literal 'Z', and a final alphanumeric
/// check-digit computed with a public mod-36 algorithm.
///
/// This validates the state code range and the embedded PAN's format (the
/// same structural check <see cref="PanValidator"/> applies), but
/// deliberately does NOT recompute the mod-36 check digit — unlike
/// Aadhaar's Verhoeff digit, getting that arithmetic subtly wrong would
/// silently reject valid GSTINs, and GSTIN is optional at Client
/// registration (Clause 3.1 of the draft Client Agreement: full business
/// verification, including GST registration, happens before a Client can
/// fund a Milestone, not as a registration gate). It also does NOT confirm
/// the GSTIN is active or matches the registering business — that needs
/// the GST Network's "Search Taxpayer" API, which belongs behind a real
/// verification integration, not here.
/// </summary>
public static partial class GstinValidator
{
    // 2 digits (state code) + PAN (5 letters, 4 digits, 1 letter) +
    // 1 entity-number char + literal 'Z' + 1 check-digit char.
    [GeneratedRegex("^[0-9]{2}[A-Z]{5}[0-9]{4}[A-Z][1-9A-Z]Z[0-9A-Z]$")]
    private static partial Regex FormatRegex();

    public static ValidationOutcome Validate(string? gstin)
    {
        if (string.IsNullOrWhiteSpace(gstin))
        {
            return ValidationOutcome.Invalid("GSTIN is required.");
        }

        var normalized = gstin.Trim().ToUpperInvariant();

        if (!FormatRegex().IsMatch(normalized))
        {
            return ValidationOutcome.Invalid(
                "GSTIN must be 15 characters: a 2-digit state code, the business's 10-character PAN, an entity number, 'Z', and a check digit.");
        }

        var stateCode = int.Parse(normalized[..2]);
        if (stateCode is < 1 or > 38)
        {
            return ValidationOutcome.Invalid("The first two digits of a GSTIN must be a valid state code (01–38).");
        }

        var embeddedPan = PanValidator.Validate(normalized.Substring(2, 10));
        if (!embeddedPan.IsValid)
        {
            return ValidationOutcome.Invalid("The PAN embedded in the GSTIN (characters 3–12) is not validly formatted.");
        }

        return ValidationOutcome.Valid;
    }

    public static string Normalize(string gstin) => gstin.Trim().ToUpperInvariant();
}
